compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -156,7 +196,12 @@ fn extract_default_variant<'a>(
         }
     };
 
-    if !matches!(variant.data, VariantData::Unit(..)) {
+    if cx.ecfg.features.default_field_values()
+        && let VariantData::Struct { fields, .. } = &variant.data
+        && fields.iter().all(|f| f.default.is_some())
+    {
+        // Allowed
```

Comment:
If we dont want to allow the following

```rust
#[derive(Default)]
enum Foo {
    #[default]
    Bar {},
}
```
we need to add an `&& !fields.is_empty()`.

Reply:
We've done this.